### DISCUSSION TITLE
Windows EnvKey cannot start anymore

### DISCUSSION THREAD

**Discussion by henrywood** [envkey/envkey#73] [Q&A]
Help !
The Windows UI program can no longer start - The load screen just stays put forever....

The last thing I did in the UI program was to click commit on a change. Upon doing this I got an error saying there was a problem with the update and from that point on I am not able to start Envkey (on Windows) any more ???

I also tried deleting the contents of %APPDATA%\envkey but that did not help...

Please help !

**Comment by henrywood** [comment]
It works now - probably something at EnvKey - Closing....

**Reply by henrywood** [reply]
Well now I does not work again ???

The problem seems to be when I commit a pending change in the Windows Envkey UI it gives an error, and then nothing is committed (it just hangs forever) but the consequence is that when I click commit, I can no longer start the Windows Envkey UI ???

Please help !

**Reply by danenania** [reply]
Hi @henrywood, we received the crash logs for this issue, but there's no obvious cause there. Could you please try the following:

- Close EnvKey UI
- Run `envkey core stop` in a terminal
- Re-open UI and try again

And let me know if that helps?

**Reply by henrywood** [reply]
Well after I did that I was able to commit my change.
But then I closed Envkey UI
And now I cannot start it again ?
Also shouldn't environment vars of a block (to which 'webserver' is connected) be written to the file, ie. when I do this in a terminal:
```bash
# envkey export webserver development /tmp/.env-webserver --format=env --cli-envkey=...
```
I would expect the vars of the block to be in /tmp/.env-webserver file along with the other ones the app has already ?
The environment vars of the block are not the /tmp/.env-webserver file ?

**Reply by henrywood** [reply]
Hi @danenania ! Is this being looked at ? It is just that EnvKey is pretty useless without the UI, so please help :)

**Reply by danenania** [reply]
When you say you can't start the UI again, what is happening? It silently fails to start or there is some kind of error dialog?

`envkey export` currently only accepts an app, though it would definitely be nice if it could take a block as well.

**Reply by henrywood** [reply]
It turns out I had two envkey processes (found by ps aux) so once I killed them I can start the UI again (and that only starts one process then).
So that part is solved, but what is the point of blocks then ?
I have a single ENV VAR (called SESSION_DURATION_SECS) in a block COMMON.
My two apps are connected to this block COMMON.
What I need is to get the environment variables that are related to the app called webserver (both the ones that come from the common block and the other ones (currently only has TEST=1)
On my servers I need to get both into the same .env file ?

**Reply by danenania** [reply]
Also, what version of Windows are you on @henrywood?

**Reply by henrywood** [reply]
@danenania I am running Windows 11

**Reply by danenania** [reply]
Ok, something must have gone wrong for two envkey processes to be running, so we'll look into that. It makes sense that it would cause issues.

It sounds like you want to connect servers to your apps? You can follow the steps here to generate an ENVKEY and load both your block and app vars: https://docs-v2.envkey.com/docs/integration-quickstart

Hope that helps?

**Reply by henrywood** [reply]
Hi again @danenania !
Well I found kind of a workaround (well I think it was your suggestion actually - I just took some time getting it...)

If I make one server key for each of my apps (I will have around 20 apps eventually) for each environment (=60 server keys), drop using envkey cli's export and just do this:
```bash
# ENVKEY=... envkey-source --dot-env > /tmp/.env-webserver
```
Then I get the env file I need !

Thanks for all your help !

**Reply by danenania** [reply]
Great, that sounds like a good approach! 

While the 'default' path described in the quickstart is to use envkey-source to pass your EnvKey variables through to your app as environment variables (either with `envkey-source -- app-start-command` or `eval $(envkey-source)` -- the latter sets env vars in the current shell), putting them in a file as you're doing can be a good solution too. That's why envkey-source includes file export options 😄 

`envkey-source` is designed to be lighter-weight and faster than the `envkey` CLI for loading an environment's config as well as providing more redundancy and robustness (multiple backups, retry logic, etc.). So it definitely sounds like the right choice for your needs.

Since you have a lot of server keys to generate, you could look into scripting their creation with the `envkey` CLI if you want to save some clicks generating them all in the UI. You can use the `envkey servers create` command for that.

I'm glad I could help! Please let me know if you run into any other issues or have more questions.

**Reply by henrywood** [reply]
@danenania henrywood here again.
Can you please show the exact steps to use envkey servers create ?

```bash
$ envkey servers create
```
gives
Authentication required. Use `envkey sign-in`, or `envkey accept-invite`
So I tried:
```bash
$ envkey sign-in
```
Gives:
No EnvKey accounts are stored on this device.

Try `envkey accept-invite`

So we try that

```bash
$ envkey accept-invite
```
It then asks me for a token (which I don't have)

In the .envkey I noticed an appId but since I have approx. 20 apps to create server keys for (and 3 environments), I don't think that is the way to go (having to provide 20 appIds / .envkey files ???)

Please help !

**Reply by danenania** [reply]
Are you running the CLI on a different server/device than the one that you installed the UI on? Or running it within a container or VM?

If that's the case, you would either need to:

- Authorize a new device on the machine where you've already installed the CLI (by clicking the `Add +` button in the sidebar on the left, then `Authorize Device`), then accept the invitation on the other machine. You would do this if you want to run the CLI on another development machine separate from the one you created your organization on.

- Create a CLI key (`Add +` > `Create CLI Key`) and then pass it through the `--cli-envkey` parameter to your CLI commands. You would do this if you want to run CLI commands on a server or in a VM/container where installing the UI doesn't make sense.

If you use the `--cli-key` parameter or set a `CLI_ENVKEY` environment variable, the CLI runs in 'Auto mode' rather than 'Interactive mode' (more on that here: https://docs-v2.envkey.com/docs/cli-overview#interactive-mode), meaning you have to provide all arguments vs. being prompted for missing arguments in interactive mode. So if you're using a CLI key, the full command would be something like:

`envkey servers create my-app production my-server-name --cli-envkey your-cli-key`

**Reply by henrywood** [reply]
I am running on the same machine that runs the Windows UI, but I am running inside WSL2 ?

**Reply by danenania** [reply]
Ok, EnvKey account info is stored in `$HOME/.envkey` so if that points to a different location under WSL, you would either need to treat it like a separate device and follow the device authorization flow I described, or else use a CLI key. Another option could be to use  git bash or CMD for your shell, where the CLI should also work.

**Reply by henrywood** [reply]
Thank you - I will look into it tomorrow - It's getting late here !
Sorry for the late reply !

**Comment by henrywood** [comment]
I now tried servers create
I did:
```bash
# envkey --cli-envkey=... seqgenerator development
```
And I got:
{"ok":false,"error":"Environment not found"}

What am I doing wrong ?

Please help !

**Reply by danenania** [reply]
Can you paste the full command (except for the `--cli-envkey`)? The following should work:

```
envkey --cli-envkey=... servers create seqgenerator development server-name
```

unless you've removed or changed the name of the development environment? You can use this to check on the environments for that app:

```
envkey --cli-envkey=... envs seqgenerator
```

**Reply by henrywood** [reply]
Hi @danenania 
Here are my findings:

```bash
# envkey --cli-envkey=... envs seqgenerator
``` 
gives:
```bash
{"ok":true,"environments":[{"id":"992e83e5-9722-49ee-9ec2-cda05f1818cf","name":"Development","branches":[]},{"id":"e7cb7e61-3879-44b1-a439-cb95f9d8a3ed","name":"Staging","branches":[]},{"id":"7be3d1b9-ed71-4cbf-8efe-f2ea0a121023","name":"Production","branches":[]}]}
```

So I thought this would work:
```bash
# envkey --cli-envkey=... servers create SEQGENERATOR seqgenerator Development
```
but that gives:
```bash
{"ok":false,"error":"Environment not found"}
```
And it does not matter if I use "Development" or "development"

Sorry for not getting back to you sooner - I was busy today :)

**Reply by danenania** [reply]
It looks like you have the environment and server name switched. Try:

```
envkey --cli-envkey=... servers create SEQGENERATOR Development seqgenerator
```

You can get help for parameters/parameter order with `envkey servers create -h`

**Reply by henrywood** [reply]
Thank you - that worked :)

**Reply by danenania** [reply]
Great!
